cudax/include/cuda/experimental/__container/async_mdarray.cuh

```diff
@@ -0,0 +1,901 @@
+  //! from \p __value
+  //! @param __mr The memory resource to allocate the async_mdarray with.
+  //! @param __size The size of the async_mdarray.
+  //! @param __value The value all elements are copied from.
+  //! @note If `__size == 0` then no memory is allocated.
+  _CCCL_HIDE_FROM_ABI explicit async_mdarray(const __env_t& __env, const size_type __size, const _Tp& __value)
+      : async_mdarray(__env, __size, ::cuda::experimental::uninit)
+  {
+    if (__size != 0)
+    {
+      this->__fill_n(__unwrapped_begin(), __size, __value);
+    }
+  }
+
+  //! @brief Constructs a async_mdarray of size \p __size using a memory and leaves all elements uninitialized
+  //! @param __mr The memory resource to allocate the async_mdarray with.
+  //! @param __size The size of the async_mdarray.
+  //! @warning This constructor does *NOT* initialize any elements. It is the user's responsibility to ensure that the
+  //! elements within `[vec.begin(), vec.end())` are properly initialized, e.g with `cuda::std::uninitialized_copy`.
+  //! At the destruction of the \c async_mdarray all elements in the range `[vec.begin(), vec.end())` will be destroyed.
+  _CCCL_HIDE_FROM_ABI explicit async_mdarray(
+    const __env_t& __env, const size_type __size, ::cuda::experimental::uninit_t)
+      : __buf_(
+          __env.query(::cuda::experimental::get_memory_resource), __env.query(::cuda::experimental::get_stream), __size)
+      , __mapping_(_Extents{__size})
+      , __policy_(__env.query(::cuda::experimental::execution::get_execution_policy))
+  {}
+
+  //! @brief Constructs a async_mdarray using a memory resource and copy-constructs all elements from the forward range
+  //! ``[__first, __last)``
+  //! @param __mr The memory resource to allocate the async_mdarray with.
+  //! @param __first The start of the input sequence.
+  //! @param __last The end of the input sequence.
+  //! @note If `__first == __last` then no memory is allocated
+  _CCCL_TEMPLATE(class _Iter)
+  _CCCL_REQUIRES(_CUDA_VSTD::__is_cpp17_forward_iterator<_Iter>::value)
+  _CCCL_HIDE_FROM_ABI async_mdarray(const __env_t& __env, _Iter __first, _Iter __last)
+      : async_mdarray(__env, static_cast<size_type>(_CUDA_VSTD::distance(__first, __last)), ::cuda::experimental::uninit)
+  {
+    if (size() > 0)
+    {
+      this->__copy_cross<_Iter, __detect_transfer_kind<__is_host_only, _Iter>>(
+        __first, __last, __unwrapped_begin(), size());
+    }
+  }
+
+  //! @brief Constructs a async_mdarray using a memory resource and copy-constructs all elements from \p __ilist
+  //! @param __mr The memory resource to allocate the async_mdarray with.
+  //! @param __ilist The initializer_list being copied into the async_mdarray.
+  //! @note If `__ilist.size() == 0` then no memory is allocated
+  _CCCL_HIDE_FROM_ABI async_mdarray(const __env_t& __env, _CUDA_VSTD::initializer_list<_Tp> __ilist)
```

Comment:
If we want construction from an `initializer_list`, should we consider construction from a `rank()`-nested `initializer_list`, so that 

a. users don't have to figure out what a flat `initializer_list<T>` means for a multidimensional array (e.g., does its order depend on the layout?);

b. users can express multidimensional arrays directly; and

c. we could provide a deduction guide that deduces the rank automatically?

Please see the relevant section of P3308R0, ["Consider adding construction from nested initializer list,"](https://isocpp.org/files/papers/P3308R0.html#consider-adding-construction-from-nested-initializer-list), for details.

Reply:
I am a bit doubtfull we want that before it has made its way into the standard. 

Also I am sceptical if this is possible with dynamic extents due to potential conflicts with list initialization of the stored type.

On the other hand direct-list initialization like here is super useful on its own, so I would definitely want to keep that in